### ISSUE TITLE
encrypt/decrypt functions desired.  Can be weak encryption.

### ISSUE THREAD

**Issue by adventuremagic123** [RPTools/maptool#974]
**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is. Ex. I'm always frustrated when 

As a GM, I would like to add encrypted notes to tokens so that if players use the Token Property Editor, they can't read the notes.

**Describe the solution you'd like**

I'd like to have new functions added, so that we could call them like:

[h:gibberish = encrypt(key, somestring)]
[h:readable = dycrypt(key, gibberish)]

**Describe alternatives you've considered**

I've considered using the GM Notes feature of the Token Property Editor, but there are reasons that doesn't work:

1. I want to be able to associate notes with other types of objects other than character tokens.  For example, equipment items, history, background, conditions, etc.

2. Using the Token Property Editor creates a concurrency issue with two or more people trying to update the token.  One of them will lose data.

**Additional context**

**Comment by Azhrei** [comment]
Well, I see two solutions right away.  Put all of your notes on a lib token that players don’t have access to (a JSON property with the token name or GUID as key).  They can’t access it at all so only trusted functions can, and if those functions start with an `isGM()` check, you’re protected.

Leave the notes on individual tokens and turn on Strict Token Ownership in the server, then don’t give players ownership of said tokens.  This doesn’t work well for NPCs when you want players to have control over those NPCs (such as henchmen or summoned monsters).

If you really want obfuscated text, a UDF can use regular expressions to do a simple textual replacement.  You know how it works:  every A is replaced with N, every B is replaced with X, and so on.  The translation would be kind of bulky code, though, I think.  There’s no way to do that with a single macro call, IIRC. 🙁

**Comment by JamzTheMan** [comment]
FWIW it would also be useful to have a "Vault" and/or encrypt/decrypt functions for other reasons.

For instance, using REST macros to call API's sometimes requires a auth token (Syrincscape and GitHub) or you may have a WebHook URL that should be secret (like Discord).

My current workaround is by hosting a php that uses oppenssl_encrypt/decrypt and call it using REST API's (along with storing in environment variables) so I can quickly wipe/restore my keys before sharing my campaign.

It still has risk though of accidentally sharing said keys.  As MT gets more connected, I see some use in this.

**Comment by adventuremagic123** [comment]
I want to use it the same way as encode/decode ... on properties of JSON objects.  Just added as a clarification.

**Comment by adventuremagic123** [comment]
I don't know if this feature, done in a far more sophisticated way, could be a doorway for commercial content -- where consumers could buy MapTool campaigns created by the likes of Paizo or other third parties.

**Comment by adventuremagic123** [comment]
> Well, I see two solutions right away. Put all of your notes on a lib token that players don’t have access to (a JSON property with the token name or GUID as key). They can’t access it at all so only trusted functions can, and if those functions start with an `isGM()` check, you’re protected.
> 
> Leave the notes on individual tokens and turn on Strict Token Ownership in the server, then don’t give players ownership of said tokens. This doesn’t work well for NPCs when you want players to have control over those NPCs (such as henchmen or summoned monsters).
> 
> If you really want obfuscated text, a UDF can use regular expressions to do a simple textual replacement. You know how it works: every A is replaced with N, every B is replaced with X, and so on. The translation would be kind of bulky code, though, I think. There’s no way to do that with a single macro call, IIRC.

Yes, I've decided to use a lib:token as a stop-gap measure.  Being able to maintain them directly with the token would be preferred.

**Comment by JamzTheMan** [comment]
On a side note, there was a similar request to "scamble" or "encrypt" text (mostly in chat IIRC) to players that had PC's that didn't know the language.

For instance, player 1 doesn't speak Orc but player 2 & 3 do. GM types in selected language "orc" and only player 2 & 3 can see it, player 2 gets gobbly gook.

Most likely separate requests and code changes but something to keep in mind if someone picks this issue up.

**Comment by adventuremagic123** [comment]
Yes, I have thought of a work-around for this issue.  Jamz suggested using base64.encode() and I've thought of that -- but it's too easy to decode and I've had players that would be tempted to hack it to investigate GM notes (or at least players who've claimed other player are bragging about things like that -- but I think that was a case of an overly suspicious player).

Anyway, as a work-around, what I can do is salt with the beginning, end, or middle of the base64.encode()'ed string with garbage that looks valid but causes a decode error.  The number of characters I use in the salt would be a fixed number that only I would know.

**Comment by adventuremagic123** [comment]
BTW, the new "Lock Player Token Editor" feature comes pretty close to resolving this issue.  If I can maintain that locked all the time, then it would be harder for players to look at and modify token properties.  They'd have to write a macro to list the available properties, another to read the properties, and another to change the properties.  Many players don't have a clue how to do this.
